### PR TITLE
Reference implementations for softmax, log_softmax, logsumexp

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This PR adds references for:

- `torch.softmax`
- `torch.log_softmax`
- `torch.logsumexp`

Unfortunately, none of them currently pass `test_python_ref_executor` even with `"aten"` executor.

### CODE REVIEW

**Comment by ngimel** [comment]
What are the errors with aten executor?

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -520,6 +520,50 @@ def log10(a):
...
+@out_wrapper
+def log_softmax(
+    a: TensorLikeType,
+    dim: int,
+    *,
+    dtype: Optional[torch.dtype] = None,
+) -> TensorLikeType:
+    result_dtype = dtype or a.dtype
+    computation_dtype = utils.get_computation_dtype(a.dtype)
+    a = prims.convert_element_type(a, computation_dtype)
```
Comment (mruberry) [inline torch/_refs/__init__.py:532]: Let's make this conditional on the conversion being required -- we actually have a function for this if you'd prefer, but it's in an awkward place (it could be moved to utils if you'd rather use it vs. a custom conditional) 

https://github.com/pytorch/pytorch/blob/38e717dc876e0f9af66b32e9cc59b44b1f9edf27/torch/_prims/wrappers.py#L20

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -520,6 +520,50 @@ def log10(a):
...
+    a_max = amax(a, dim, keepdim=True)
+    shifted = a - a_max
```
Comment (mruberry) [inline torch/_refs/__init__.py:534]: Nit: comment for why the shift occurs would be nice

File: torch/_refs/__init__.py [hunk 3]
```diff
@@ -520,6 +520,50 @@ def log10(a):
...
+    shifted_logsumexp = logsumexp(shifted, dim, keepdim=True)
+    return prims.convert_element_type(
```
Comment (mruberry) [inline torch/_refs/__init__.py:536]: We should also make this conditional on the conversion being required

File: torch/_refs/__init__.py [hunk 4]
```diff
@@ -520,6 +520,50 @@ def log10(a):
...
+        shifted - shifted_logsumexp, result_dtype
+    )
+
+
+def _squeeze_multiple(a: TensorLikeType, dims: DimsSequenceType) -> TensorLikeType:
```
Comment (mruberry) [inline torch/_refs/__init__.py:541]: Unify this with prims.squeeze -- see 

https://github.com/pytorch/pytorch/blob/38e717dc876e0f9af66b32e9cc59b44b1f9edf27/torch/_prims/__init__.py#L1636

I think we only need one? But maybe this implementation is better for the prim?

File: torch/_refs/__init__.py [hunk 5]
```diff
@@ -520,6 +520,50 @@ def log10(a):
...
+    for dim in reversed(dims):
+        a = squeeze(a, dim)
+    return a
+
+
+@out_wrapper
+def logsumexp(
+    a: TensorLikeType,
+    dims: DimsType,
```
Comment (mruberry) [inline torch/_refs/__init__.py:550]: "dim"? 

https://pytorch.org/docs/master/generated/torch.logsumexp.html?highlight=logsumexp#torch.logsumexp

File: torch/_refs/__init__.py [hunk 6]
```diff
@@ -520,6 +520,50 @@ def log10(a):
...
+    keepdim: bool = False,
+) -> TensorLikeType:
+    dims = utils.canonicalize_dims(a.ndim, dims)
+    # ATen specifies int[1] type dims which expands integers to tuples of length 1
```
Comment (mruberry) [inline torch/_refs/__init__.py:554]: Nice comment

File: torch/_refs/__init__.py [hunk 7]
```diff
@@ -520,6 +520,50 @@ def log10(a):
...
+    if not isinstance(dims, Iterable):
+        dims = (dims,)
+    if utils.is_float_dtype(a.dtype) or utils.is_complex_dtype(a.dtype):
+        a_max = amax(a, dims, keepdim=True)
+        a_max = where(abs(a_max) == float("inf"), 0.0, a_max)
+        a_max_squeezed = _squeeze_multiple(a_max, dims) if not keepdim else a_max
+        result = log(sum(exp(a - a_max), dims, keepdim=keepdim)) + a_max_squeezed
+    else:
+        result = log(sum(exp(a), dims, keepdim=keepdim))
```
Comment (mruberry) [inline torch/_refs/__init__.py:563]: Add a comment for what this case covers (integer and boolean dtypes)

File: torch/_refs/__init__.py [hunk 8]
```diff
@@ -520,6 +520,50 @@ def log10(a):
...
+    dims: DimsType,
+    keepdim: bool = False,
+) -> TensorLikeType:
+    dims = utils.canonicalize_dims(a.ndim, dims)
+    # ATen specifies int[1] type dims which expands integers to tuples of length 1
+    if not isinstance(dims, Iterable):
+        dims = (dims,)
+    if utils.is_float_dtype(a.dtype) or utils.is_complex_dtype(a.dtype):
+        a_max = amax(a, dims, keepdim=True)
+        a_max = where(abs(a_max) == float("inf"), 0.0, a_max)
```
Comment (mruberry) [inline torch/_refs/__init__.py:559]: really elegant code here

File: torch/_refs/__init__.py [hunk 9]
```diff
@@ -2017,6 +2061,23 @@ def stack(tensors: TensorSequenceType, dim: int = 0) -> TensorLikeType:
...
+@out_wrapper
+def softmax(
+    a: TensorLikeType,
+    dim: int,
+    *,
+    dtype: Optional[torch.dtype] = None,
+) -> TensorLikeType:
+    result_dtype = dtype or a.dtype
+    computation_dtype = utils.get_computation_dtype(a.dtype)
+    a = prims.convert_element_type(a, computation_dtype)
```
Comment (mruberry) [inline torch/_refs/__init__.py:2073]: Similar comments here as with log_softmax re: conditional conversions

**Review by mruberry — APPROVED** [review]
This is awesome, @IvanYashchuk! I made some inline comments for your review, nothing major, and the lint job needs to be fixed, but approving this for velocity because I'm sure you'll sort out the review and the jobs.

File: torch/_refs/__init__.py (hunk 5 above)
Comment (IvanYashchuk) [reply torch/_refs/__init__.py:550]: Right, it should be dim to match the torch namespace. I think I was confused because it also accepts and works for several dimensions
> dim (int or tuple of python:ints)
```py
import torch
a = torch.ones(3, 3)
torch.logsumexp(a, (0, 1))
# tensor(3.1972)
```

File: torch/_refs/__init__.py (hunk 2 above)
Comment (IvanYashchuk) [reply torch/_refs/__init__.py:534]: Oh, the shift is actually not required here because stabilized logsumexp is used. I will remove it.

File: torch/_refs/__init__.py (hunk 1 above)
Comment (IvanYashchuk) [reply torch/_refs/__init__.py:532]: Thanks! `_maybe_convert_to_dtype` is already used in this file, so we can keep it defined in the awkward place for now 🙂

File: torch/_refs/__init__.py (hunk 4 above)
Comment (IvanYashchuk) [reply torch/_refs/__init__.py:541]: I didn't notice that `prims.squeeze` works for multiple specified dimensions. Both implementations are fine, I will not touch `prims.squeeze` at this time.

**Comment by IvanYashchuk** [comment]
`_maybe_convert_to_dtype` is now used for optional conversions. I removed unnecessary shift from log_softmax and fixed mypy errors. In case when `_maybe_convert_to_dtype` is the returned variable mypy errors are silenced.
